packages/api/src/command/medical/cq-directory/rebuild-cq-directory.ts

```diff
@@ -0,0 +1,105 @@
+import { Config } from "../../../shared/config";
+import { capture } from "../../../shared/notifications";
+import { bulkInsertCQDirectoryEntries } from "./create-cq-directory-entry";
+import { parseCQDirectoryEntries } from "./parse-cq-directory-entry";
+
+const BATCH_SIZE = 1000;
+const cqDirectoryEntryTemp = `cq_directory_entry_temp`;
+const cqDirectoryEntry = `cq_directory_entry`;
+const cqDirectoryEntryBackup = `cq_directory_entry_backup`;
+
+const sqlDBCreds = Config.getDBCreds();
+const dbCreds = JSON.parse(sqlDBCreds);
+
+const sequelize = new Sequelize(dbCreds.dbname, dbCreds.username, dbCreds.password, {
+  host: dbCreds.host,
+  port: dbCreds.port,
+  dialect: dbCreds.engine,
+});
+
+export const rebuildCQDirectory = async (failGracefully = false): Promise<void> => {
+  let currentPosition = 0;
+  let isDone = false;
+
+  try {
+    await createTempCQDirectoryTable();
+    while (!isDone) {
+      try {
+        const apiKey = Config.getCQApiKey();
+        const cq = new Carequality(apiKey);
+        const orgs = await cq.listOrganizations({ start: currentPosition });
+        if (orgs.length < BATCH_SIZE) isDone = true; // if CQ directory returns less than BATCH_SIZE number of orgs, that means we've hit the end
+        currentPosition += BATCH_SIZE;
+        const parsedOrgs = parseCQDirectoryEntries(orgs);
+        await bulkInsertCQDirectoryEntries(sequelize, parsedOrgs);
+        console.log(`Added ${currentPosition} CQ directory entries...`);
+        await sleep(750);
+      } catch (error) {
+        isDone = true;
+        if (!failGracefully) {
+          throw error;
+        }
+      }
+    }
+  } catch (error) {
+    await deleteTempCQDirectoryTable();
+    const msg = `Failed to rebuild the directory`;
+    console.log(`${msg}, error: ${error}`);
+    capture.message(msg, {
+      extra: { context: `rebuildCQDirectory`, error },
+      level: "error",
+    });
+    throw error;
+  }
+  try {
+    await renameCQDirectoryTablesAndUpdateIndexes();
+  } catch (error) {
+    const msg = `Failed the last step of CQ directory rebuild`;
+    console.log(`${msg}. Cause: ${error}`);
+    capture.message(msg, {
+      extra: { context: `renameCQDirectoryTablesAndUpdateIndexes`, error },
+      level: "error",
+    });
+    throw error;
+  }
+};
+
+const createTempCQDirectoryTable = async () => {
+  const query = `CREATE TABLE IF NOT EXISTS ${cqDirectoryEntryTemp} AS SELECT * FROM ${cqDirectoryEntry} WHERE 1=0;`;
+  await sequelize.query(query, {
+    type: QueryTypes.INSERT,
+  });
+};
+
+const deleteTempCQDirectoryTable = async () => {
+  const query = `DROP TABLE IF EXISTS ${cqDirectoryEntryTemp}`;
+  await sequelize.query(query, {
+    type: QueryTypes.DELETE,
+  });
+};
+
+const renameCQDirectoryTablesAndUpdateIndexes = async () => {
+  const dropBackupQuery = `DROP TABLE IF EXISTS ${cqDirectoryEntryBackup};`;
+  await sequelize.query(dropBackupQuery, {
+    type: QueryTypes.DELETE,
+    logging: false,
+  });
```

Comment:
So we're not keeping a history of the last few executions, is that correct? I thought we wanted to, but NABD at this point I guess.

Reply:
Yea, I think for now, just one copy is probably good enough. If down the line, there's a need for it, we can update it